### PR TITLE
Add optional `ray_distances` property to `RayCasterData`

### LINKED ISSUE #2946: [Proposal] Add `ray_distances` for RayCasterData
### Proposal
Add a `ray_distances` field to the RayCasterData output to directly provide Euclidean distances for all rays.

### Motivation
Current implementation forces users to manually calculate distances using `ray_hits_w` and `pos_w` (e.g., `torch.norm(ray_hits_w - pos_w, axis=-1)`). This requires redundant boilerplate code in every RL environment implementation and introduces unnecessary complexity and potential errors for common tasks like collision avoidance.

### Alternatives
Considered providing a helper function, but direct integration offers better API consistency and eliminates redundant user-side calculations.

### Implementation
Simple interface modification using existing raycast data. Requires minimal changes to compute distances from `ray_hits_w` and `pos_w`. For unhit rays, returns `max-distance` to match existing convention.

### Additional context
Fundamental for RL tasks like navigation and object avoidance. Eliminates common pain point reported by users.

### Checklist
- [x] I have checked that there is no similar issue in the repo

### Acceptance Criteria
- [ ] `ray_distances` field added to RayCasterData output
- [ ] Documentation updated across relevant files

### PR DESCRIPTION
**Pull request by taceris**
# Description

This PR adds an optional `ray_distances` property to `RayCasterData` to directly provide Euclidean distances for all rays. This eliminates the need for users to manually calculate distances in every implementation, reducing boilerplate code and potential errors in RL environments.

Key changes:
1. Added `ray_distances` property from existing raycast data
2. For unhit rays, returns `cfg.max_distance` to match existing conventions

Fixes https://github.com/isaac-sim/IsaacLab/issues/2946

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added my name to the `CONTRIBUTORS.md`

### CODE REVIEW

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_cfg.py [hunk 1]
```diff
@@ -62,6 +61,9 @@ class OffsetCfg:
         * `world` if rays' starting positions and directions are always fixed. This is useful in combination with the grid map package.
     """
 
+    return_distance: bool = False
```
Comment (Mayankm96) [inline source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_cfg.py]: Can we call it `track_ray_distance`. This will then match the way we handle different data outputs in ContactSensor class:
https://github.com/isaac-sim/IsaacLab/blob/main/source/isaaclab/isaaclab/sensors/contact_sensor/contact_sensor_cfg.py#L20-L24

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py [hunk 2]
```diff
@@ -27,3 +27,9 @@ class RayCasterData:
     Shape is (N, B, 3), where N is the number of sensors, B is the number of rays
     in the scan pattern per sensor.
     """
+    ray_distance: torch.Tensor = None
```
Comment (Mayankm96) [inline source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py]: Please follow documentation as here: https://github.com/isaac-sim/IsaacLab/blob/main/source/isaaclab/isaaclab/sensors/contact_sensor/contact_sensor_data.py#L89

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_cfg.py (hunk 1 above)
Comment (taceris) [reply source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_cfg.py]: Ok, the current format has been matched with the ContactSensor class now.

>     track_ray_distance: bool = False
>     """Whether to track the ray_distance (distances of the rays until they hits the mesh). Defaults to False."""

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py (hunk 2 above)
Comment (taceris) [reply source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py]: Done.

>     ray_distance: torch.Tensor | None = None
>     """Distances (in meters) of the rays until they hits the mesh. Clipped to 'RayCasterCfg.max_distance'.
> 
>     Shape is (N, B), where N is the number of sensors, B is the number of rays
>     in the scan pattern per sensor.
> 
>     Note:
>         If the :attr:`RayCasterCfg.track_ray_distance` is False, then this quantity is None.
>     """

**Review by Mayankm96 — APPROVED** [review]

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_cfg.py [hunk 3]
```diff
@@ -62,6 +61,9 @@ class OffsetCfg:
         * `world` if rays' starting positions and directions are always fixed. This is useful in combination with the grid map package.
     """
 
+    track_ray_distance: bool = False
+    """Whether to track the ray_distance (distances of the rays until they hits the mesh). Defaults to False."""
```
Comment (Mayankm96) [inline source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_cfg.py]:
```suggestion
    """Whether to track the raycast distances (i.e., the distances traveled by the rays until they hit a target mesh). Defaults to False."""
```

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster.py [hunk 4]
```diff
@@ -226,6 +226,8 @@ def _initialize_rays_impl(self):
         self._data.pos_w = torch.zeros(self._view.count, 3, device=self._device)
         self._data.quat_w = torch.zeros(self._view.count, 4, device=self._device)
         self._data.ray_hits_w = torch.zeros(self._view.count, self.num_rays, 3, device=self._device)
+        if self.cfg.track_ray_distance:
+            self._data.ray_distance = torch.ones(self._view.count, self.num_rays, device=self._device)
```
Comment (Mayankm96) [inline source/isaaclab/isaaclab/sensors/ray_caster/ray_caster.py]:
```suggestion
            self._data.ray_distance = torch.full((self._view.count, self.num_rays), self.cfg.max_distance, device=self._device)
```

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py [hunk 5]
```diff
@@ -27,3 +27,12 @@ class RayCasterData:
     Shape is (N, B, 3), where N is the number of sensors, B is the number of rays
     in the scan pattern per sensor.
     """
+    ray_distance: torch.Tensor | None = None
+    """Distances (in meters) of the rays until they hits the mesh. Clipped to 'RayCasterCfg.max_distance'.
+
+    Shape is (N, B), where N is the number of sensors, B is the number of rays
```
Comment (Mayankm96) [inline source/isaaclab/isaaclab/sensors/ray_caster/ray_caster_data.py]:
```suggestion
    """Distances (in meters) of the rays until they hit a target mesh.
    
    This distance is clipped to the maximum distance specified through
    :attr:`RayCasterCfg.max_distance`.

    Shape is (N, B), where N is the number of sensors, B is the number of rays
```

**Comment by Mayankm96** [comment]
@raidenLight I think we are missing a test for this feature. Could we have that too in the MR? Also need to update the CHANGELOG and extension.toml file.

Thanks again!

**Comment by ooctipus** [comment]
Hi, thanks @raidenLight for the PR,  Are you willing add test case for this new feature? Let me know if you need help : ))

**Comment by taceris** [comment]
> @raidenLight I think we are missing a test for this feature. Could we have that too in the MR? Also need to update the CHANGELOG and extension.toml file.
> 
> Thanks again!


> Hi, thanks @raidenLight for the PR, Are you willing add test case for this new feature? Let me know if you need help : ))


@Mayankm96 @ooctipus Thank you for your suggestions! This is my first time using git-pull, and I'm not entirely sure how to add tests (is it something like source/isaaclab/test/sensors/check_ray_caster.py?), or how to update the CHANGELOG and extension.toml file. I'll be very glad if you could give me a brief explanation :)

**Comment by ooctipus** [comment]
@raidenLight
Congratulation on your first PR! 

Ahh, I just check, since we don't have ray_caster test yet, and this is your first time starting pull request, I think it is a bit too much to ask you to come with tests on your own. I think what will be nice instead will be printing the ray_distances in

`scripts/tutorials/04_sensors/run_ray_caster.py`
after you configured your ray caster to report ray distance
in below block add the printing, (please verify that the value makes sense to you):
```
with Timer(
            f"Ray-caster update with {4} x {ray_caster.num_rays} rays with max height of"
            f" {torch.max(ray_caster.data.pos_w).item():.2f}"  #  ---------- Add your ray_distance print
        ):
        
```

For the changelog, since raycaster.py is isaaclab.sensor.RayCaster, you would want to add a new entry in `source/isaaclab/docs/CHANGELOG.rst,` following the pattern of other entries.

Then increment the package version in source/isaaclab/config/extension.toml to the number you put in CHANGELOG

: ))
Let me know if you have any questions.

**Comment by taceris** [comment]
> @raidenLight Congratulation on your first PR!恭喜您提交了第一个 PR！
> 
> Ahh, I just check, since we don't have ray_caster test yet, and this is your first time starting pull request, I think it is a bit too much to ask you to come with tests on your own. I think what will be nice instead will be printing the ray_distances in啊，我刚检查了一下，由于我们还没有 ray_caster 测试，而且这是你第一次提交 pull request，我认为让你自己写测试有点太多了。我觉得更好的做法是在你配置了 ray caster 以报告 ray distance 后，在
> 
> `scripts/tutorials/04_sensors/run_ray_caster.py` after you configured your ray caster to report ray distance中打印 ray_distances。 in below block add the printing, (please verify that the value makes sense to you):在下面的代码块中添加打印功能，请确认这些值对你来说是有意义的：
> 
> ```
> with Timer(
>             f"Ray-caster update with {4} x {ray_caster.num_rays} rays with max height of"
>             f" {torch.max(ray_caster.data.pos_w).item():.2f}"  #  ---------- Add your ray_distance print
>         ):
>         
> ```
> 
> For the changelog, since raycaster.py is isaaclab.sensor.RayCaster, you would want to add a new entry in `source/isaaclab/docs/CHANGELOG.rst,` following the pattern of other entries.对于 changelog，由于 raycaster.py 是 isaaclab.sensor.RayCaster，你需要在 `source/isaaclab/docs/CHANGELOG.rst,` 中按照其他条目的模式添加一个新条目。
> 
> Then increment the package version in source/isaaclab/config/extension.toml to the number you put in CHANGELOG然后，将 source/isaaclab/config/extension.toml 中的包版本增加到你在 CHANGELOG 中填写的数字
> 
> : )) Let me know if you have any questions.如果你有任何问题，请告诉我。

@ooctipus Thank you very much for your patient help! 

I have added a simple output of `ray_distance` in `scripts/tutorials/04_sensors/run_ray_caster.py`. 
Snipaste_2025-08-07_02-20-33

Additionally, I created a new simple example `scripts/tutorials/04_sensors/run_ray_caster_on_crazyfile.py` to more clearly demonstrate how the newly added property can assist with robot navigation tasks. 
Snipaste_2025-08-07_02-17-06

I'm not sure if it's necessary, but I hope it can be helpful.

: ))

File: source/isaaclab/isaaclab/sensors/ray_caster/ray_caster.py (hunk 4 above)
Comment (taceris) [reply source/isaaclab/isaaclab/sensors/ray_caster/ray_caster.py]: > self._data.ray_distance = torch.full((self._view.count, self.num_rays), self.cfg.max_distance, device=self._device)

It seems that explicitly specifying the data type would be better,  `self._data.ray_distance = torch.full((self._view.count, self.num_rays), self.cfg.max_distance, device=self._device, dtype=torch.float32)`, otherwise it might cause an error if the user sets max_distance as an integer in the config (e.g., max_distance=5).
